### ISSUE TITLE
Is hessian gain supported or not? on what loss?

### ISSUE THREAD

**Issue by Willian-Zhang** [google/yggdrasil-decision-forests#12]
Yggdrasil is such a nice project as it has such versatile support on GBT. 
One intriguing is that Yggdrasil supports  `use_hessian_gain ` as a toggle for different training style.

I'm currently looking into implementing my own Loss function for certain tasks, yet question raises when dealing with `Hessian`.
- Is hessian essential for a new loss func implemented 
- Is use_hessian_gain behavior actually defaulted to what's set in Loss, or really just from `use_hessian_gain=false` proto?
- if wanted, how to properly support hessian training
  -  Is Yggdrasil training on  `gain = G^2/(H+λ)`

## S: use_hessian_gain "Available for all losses except regression."
https://github.com/google/yggdrasil-decision-forests/blob/be943445871d61726da063c838b213d20a8104a5/documentation/learners.md#L303-L309

### A.1 . Error when use_hessian_gain = true and hessian_col_idx not set
https://github.com/google/yggdrasil-decision-forests/blob/1700e1908e04d8af9fc40c207b685786f001a541/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees.cc#L1231-L1235

### A.2 gradient.hessian_col_idx set iff hessian = true
https://github.com/google/yggdrasil-decision-forests/blob/1700e1908e04d8af9fc40c207b685786f001a541/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees.cc#L2378-L2390

### A.3 has hessian = false for MSE
https://github.com/google/yggdrasil-decision-forests/blob/15781866817e2c0d9d94ebca3e7a5fea50242678/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees_loss.h#L288-L299

### A.4 has hessian = use_hessian_gain for LogLikelihoodLoss
https://github.com/google/yggdrasil-decision-forests/blob/15781866817e2c0d9d94ebca3e7a5fea50242678/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees_loss.h#L226-L238

### B.1 Error when use_hessian_gain on Task::CLASSIFICATION
https://github.com/google/yggdrasil-decision-forests/blob/f897874b809520e1d8f8296d995de9fb8ef59671/yggdrasil_decision_forests/learner/decision_tree/training.cc#L1207-L1210

### B.2 No error when use_hessian_gain on Task::REGRESSION
https://github.com/google/yggdrasil-decision-forests/blob/f897874b809520e1d8f8296d995de9fb8ef59671/yggdrasil_decision_forests/learner/decision_tree/training.cc#L1237-L1240

### C.1 : G, H, W (sum of gradience hessian and weights) available for LogLikelihoodLoss, nothing else
https://github.com/google/yggdrasil-decision-forests/blob/15781866817e2c0d9d94ebca3e7a5fea50242678/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees_loss.cc#L283-L287

### C.2: G, H, W only accessed on Task::REGRESSION, nothing else
https://github.com/google/yggdrasil-decision-forests/blob/f897874b809520e1d8f8296d995de9fb8ef59671/yggdrasil_decision_forests/learner/decision_tree/training.cc#L1237-L1251

### D.1: MSE only available for Regression or Ranking, not Classification
https://github.com/google/yggdrasil-decision-forests/blob/15781866817e2c0d9d94ebca3e7a5fea50242678/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees_loss.cc#L369-L375

### D.2 LogLikelihood Loss only available for Classification
https://github.com/google/yggdrasil-decision-forests/blob/15781866817e2c0d9d94ebca3e7a5fea50242678/yggdrasil_decision_forests/learner/gradient_boosted_trees/gradient_boosted_trees_loss.cc#L150-L154

# Conclusion draw from S, A, B, C, D:
1.
S,B => conflicts 
either `use_hessian_gain` is documented wrong, or it could be feature not implemented yet, or there is something really wrong going on here.

2. 
C,D => hessian not trainable
as LogLikelihood Algo does not make much sense in Regression tasks, It's much likely C is invalid

3.
A, D => E: hessian available for Classification, not  Regression/Ranking
B, E => conflicts 
as B also conflicts in 1, it's likely B is the actual False one.

**Comment by achoum** [comment]
Hi Willian,

Nice analysis :).
Let me try to answer your questions.

**Background**

- `use_hessian_gain` is an hyper-parameter of the GBT learner (only i.e. it is not used in the Random Forest learner).
- By default use_hessian_gain=false for the gbdt.
- If use_hessian_gain=true, the splitting score and the leaf values are set using the equation G^2/(H+λ) (with an extra L1 regularization). For example, see LabelHessianNumericalScoreAccumulator::Score.
- If use_hessian_gain=false, only leaf values are set using a Newton step (i.e. hessian) while the score is the reduction is gradient variance. See LabelNumericalScoreAccumulator::Score.
- Hessian learning is disabled for GBT regression with MSE loss to simplify the code. Note that in this case the hessian is a constant i.e. it is equivalent to non-hessian learning.
- The GBT learning code sits in learner/gradient_boosted_trees. It relies on the generic decision forest learning code sitting in learner/decision_tree. Independently of its loss, a GBT is always composed of regression trees. For this reason, when learning a GBT, the learner/decision_tree code will always see a model::proto::Task::REGRESSION task. For example, when training a classification GBT trained with task=model::proto::Task::CLASSIFICATION and a LogLikelihood loss, the learner/decision_tree code is called with task=model::proto::Task::REGRESSION.
- To create a new loss, you need to derive and register a AbstractLoss class.

**Your questions**

>> Is hessian essential for a new loss func implemented

If gbt_config_.use_hessian_gain()=true, the hessian array should be populated by your AbstractLoss class. If gbt_config_.use_hessian_gain()=false, it is up to you (depending on what is the most efficient).

For example, in BinomialLogLikelihoodLoss, it is more efficient to recompute the hessian instead of storing it. Therefore, the hessian array is only set if gbt_config_.use_hessian_gain()=true. 

On the other hand, in NDCGLoss, the hessian is expensive to compute. Therefore, it is always store in the hessian array.

>> Is use_hessian_gain behavior actually defaulted to what's set in Loss, or really just from use_hessian_gain=false proto?

The proto. The AbstractLoss has no control over enabling or disabling the hessian logic.

>> if wanted, how to properly support hessian training

Make sure your AbstractLoss class sets the gradient and hessian arrays. When training, make sure to set the hyper-parameter use_hessian_gain to true.

I hope this helps.
